File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
def dfs(level, idx, start_team):
    global res
    if level == n // 2:
        link_team = [i for i in range(n) if i not in start_team]
        
        start_score, link_score = 0, 0
        for i in range(n // 2):
            for j in range(n // 2):
                start_score += arr[start_team[i]][start_team[j]]
                link_score += arr[link_team[i]][link_team[j]]
                
        res = min(res, abs(start_score - link_score))
        return
    
    for i in range(idx, n):
        dfs(level + 1, i + 1, start_team + [i])

n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]

res = float('inf')
dfs(0, 0, [])
print(res)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.util.*;
import java.io.*;

public class Main {
    static int n;
    static int[][] arr;
    static int min = Integer.MAX_VALUE;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        arr = new int[n][n];
        for (int i = 0; i<n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j<n; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        visited = new boolean[n];
        dfs(0, 0);
        System.out.println(min);
    }

    public static void dfs(int level, int idx) {
        if (level==n/2) {
            int startScore = 0;
            int linkScore = 0;
            for (int i = 0; i<n; i++) {
                for (int j = 0; j<n; j++) {
                    if (visited[i] && visited[j]) {
                        startScore += arr[i][j];
                    } else if (!visited[i] && !visited[j]) {
                        linkScore += arr[i][j];
                    }
                }
            }
            min = Math.min(min, Math.abs(startScore-linkScore));
            return;
        }
        for (int i = idx; i<n; i++) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(level+1, i+1);
                visited[i] = false;
            }
        }
    }
}